samples/nrf9160/mqtt_simple/src/main.c

```diff
@@ -354,18 +382,33 @@ void main(void)
 	}
 
 	while (1) {
-		err = mqtt_input(&client);
-		if (err != 0) {
-			printk("ERROR: mqtt_input %d\n", err);
+		err = poll(&fds, 1, K_SECONDS(CONFIG_MQTT_KEEPALIVE));
+		if (err < 0) {
+			printk("ERROR: poll %d\n", errno);
+			continue;
+		} else if (err == 0) {
+			printk("poll() timed out, sending ping if needed...\n");
+
+			err = mqtt_live(&client);
+			if (err != 0) {
+				printk("ERROR: mqtt_live %d\n", err);
+			}
+			continue;
 		}
 
-		err = mqtt_live(&client);
-		if (err != 0) {
-			printk("ERROR: mqtt_live %d\n", err);
+		if ((fds.revents & POLLIN) == POLLIN) {
+			err = mqtt_input(&client);
+			if (err != 0) {
+				printk("ERROR: mqtt_input %d\n", err);
+			}
 		}
 
-		if (poll(&fds, 1, K_SECONDS(CONFIG_MQTT_KEEPALIVE)) < 0) {
-			printk("ERROR: poll %d\n", errno);
+		if ((fds.revents & POLLERR) == POLLERR) {
+			printk("POLLERR\n");
```

Comment:
I'd be more strict with error processing - `POLLERR`/`POLLNVAL` indicates there's something wrong with the socket and it should no longer be used. The same would apply in case `mqtt_*` calls fail.

I'd `break` the loop in case of error, and call `mqtt_disconnect` after it.

Reply:
True, I'll add a `break` and disconnect on error.